datalad/tests/utils.py

```diff
@@ -147,58 +147,92 @@ def ok_clean_git_annex_proxy(path):
     """
     # TODO: May be let's make a method of AnnexRepo for this purpose
 
-    ar = AnnexRepo(path)
-    cwd = getpwd()
-    chpwd(path)
-
-    try:
-        out = ar.proxy(['git', 'status'])
-    except CommandNotAvailableError as e:
-        raise SkipTest
-    finally:
-        chpwd(cwd)
-
-    assert_in(
-        "nothing to commit", out[0],
-        msg="git-status output via proxy not plausible: %s" % out[0]
-    )
+    if isinstance(path, AnnexRepo):
+        ar = path
+    else:
+        ar = AnnexRepo(path)
+        # Note: Possible exception while trying to instantiate AnnexRepo
+        # intended to be thrown in order to have a failure in the assertion
+    ok_(not ar.dirty)
 
 
-def ok_clean_git(path, annex=True, head_modified=[], index_modified=[], untracked=[]):
+def ok_clean_git(path, annex=None, head_modified=[], index_modified=[],
+                 untracked=[]):
     """Verify that under given path there is a clean git repository
 
     it exists, .git exists, nothing is uncommitted/dirty/staged
+
+    Parameters
+    ----------
+    path: str or Repo
+      in case of a str: path to the repository's base dir;
+      Note, that passing a Repo instance prevents detecting annex. This might be
+      useful in case of a non-initialized annex, a GitRepo is pointing to.
+    annex: bool or None
+      explicitly set to True or False to indicate, that an annex is (not)
+      expected; set to None to autodetect, whether there is an annex.
+      Default: None.
     """
-    from datalad.support.gitrepo import GitRepo
-    if isinstance(path, GitRepo):
-        path = path.path
 
-    ok_(exists(path))
-    ok_(exists(join(path, '.git')))
-    if annex:
-        ok_(exists(join(path, '.git', 'annex')))
-    repo = git.Repo(path)
+    if isinstance(path, AnnexRepo):
+        if annex is None:
+            annex = True
+        # if `annex` was set to False, but we find an annex => fail
+        assert_is(annex, True)
+        r = path
+    elif isinstance(path, GitRepo):
+        if annex is None:
+            annex = False
+        # explicitly given GitRepo instance doesn't make sense with 'annex' True
+        assert_is(annex, False)
+        r = path
+    else:
+        # 'path' is an actual path
+        try:
+            r = AnnexRepo(path, init=False, create=False)
+            if annex is None:
+                annex = True
+            # if `annex` was set to False, but we find an annex => fail
+            assert_is(annex, True)
+        except Exception:
+            # Instantiation failed => no annex
+            r = GitRepo(path, init=False, create=False)
+            if annex is None:
+                annex = False
+            # explicitly given GitRepo instance doesn't make sense with
+            # 'annex' True
+            assert_is(annex, False)
+
+    if annex and r.is_direct_mode():
+        ok_(not r.dirty)
```

Comment:
> As for testing the existence of r.path and .git/annex: It's done ;-)

my comment was about the fact that those conditions are explicitly tested also in `else:` but not here.  Either they should be tested uniformly or not tested at all explicitly relying on those calls above as you have mentioned.  Otherwise it would be for the next developer to "figure it out" why it is so

Re absent testing in direct mode -- please at least add a big fat "TODO: " comment so at least there is some indication for future generations that we did have good intentions here but just lacked time/inspiration to do that ;)